E16n serialization (slightly more conservative version)

cwillisf opened the pull request:
This is the same as #2245 but does not move the Extension Manager into the Runtime. Instead, only the information about currently-loaded extensions is moved to the Runtime.

Commit:
move extensionManager from vm to runtime

Commit:
add extension info to block info and service name to extension info

Commit:
Add specific dispatch.callSync exception for method-not-found

Commit:
allow an extension to override block serialization

Commit:
Revert "Add specific dispatch.callSync exception for method-not-found"

This reverts commit 293e49e639c0eb6753ddd7fb71ea03728a696f6a.

Commit:
WIP e16n serialization

Commit:
support custom serialization functions for extensions

Commit:
Revert "move extensionManager from vm to runtime"

This reverts commit 8325e4b1c5046db1fdc94986791ad0294e02319a.

Commit:
move extension info map to Runtime

kchadha commented on src/extension-support/extension-manager.js:
```diff
@@ -171,16 +155,25 @@ class ExtensionManager {
                     info = this._prepareExtensionInfo(serviceName, info);
+                    loadedExtensions.set(info.id, info);
+                    if (oldId !== info.id) {
+                        log.warn(`Extension changed ID from ${oldId} to ${info.id}!`);
```
Hmmm is this still relevant now that the entire extension manager is no longer being moved to the runtime in this PR? There shouldn't really be a way to change the extension ID right?

cwillisf replied:
Note from conversation with @kchadha: it's probably best to keep `oldId` instead of switching to `info.id` here.

kchadha commented on src/blocks/data.js:
```diff
@@ -7,6 +7,10 @@ const formatMessage = require('format-message');
 
+// serialization primitives
+const VAR_PRIMITIVE = 12;
+const LIST_PRIMITIVE = 13;
```
This isn't getting used yet and is causing a lint error.

kchadha commented on src/engine/blocks.js:
```diff
@@ -547,6 +547,13 @@ class Blocks {
+        const categoryId = block.categoryId ||
+            (block.mutation && block.mutation.blockInfo && block.mutation.blockInfo.categoryId);
+        if (categoryId) {
+            block.extensionInfo = this.runtime.getExtensionInfo(categoryId);
```
I'm worried about this approach of including the entire extension info in the block vm representation. I feel like I would rather have the block just know about what category it belongs to (e.g. it already has this info in its opcode, but I could also see us adding another property called `categoryId`). 

I think instead of having the entire `extensionInfo` embedded in the block VM representation (which I believe is currently just plain JSON) we should just have the serialization code look up the extension info in the runtime when checking for whether a block has custom serialization code. See corresponding comment in the `serialization/sb3.js` code.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -95,16 +95,42 @@ const primitiveOpcodeInfoMap = {
-    if (hasOwnProperty.call(primitiveOpcodeInfoMap, block.opcode)) {
+const serializePrimitiveBlock = function (block, target) {
+    const customSerialize = getSerializationInfo(block.extensionInfo, block.opcode).serialize;
+    if (customSerialize) {
```
I would rather have this code look up whether the given block belongs to a loaded extension with extensionInfo here via `runtime.getExtensionInfo(block.categoryId)` instead of the VM block JSON representation owning the `extensionInfo` which ultimately has a serialization function nested inside it.

kchadha replied:
As far as how to get access to the runtime.... the top level serialize function has it (which ultimately calls this function through a series of nested calls). Maybe we can do something clever and figure out which blocks have custom serializers ahead of time (or we can just pass the runtime object through to this function).

I'm happy to talk more about all of this offline.

cwillisf replied:
I wasn't thinking of the block "owning" an extensionInfo -- just "borrowing" it from the extension system. That said, I do agree that keeping the VM block JSON simple (as you mentioned in the [`blocks.js` comment above](https://github.com/LLK/scratch-vm/pull/2279#discussion_r325419366)) might be a good goal. I'll try removing `block.extensionInfo` and see how it goes.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -288,17 +315,18 @@ const getExtensionIdForOpcode = function (opcode) {
  * in the serialized blocks.
  */
-const serializeBlocks = function (blocks) {
+const serializeBlocks = function (target) {
```
I think this function was set up on purpose to take just the VM block representation, so that one could have a way of serializing/deserializing just the blocks without the context of the target (to be used for other tools etc.). That's why this file exports `serializeBlocks` and `deserializeBlocks` separately alongside the full `serialize` and `deserialize` functions.

kchadha replied:
See https://github.com/LLK/scratch-vm/pull/1468

kchadha replied:
I guess this in general become more difficult to do with our custom serialization approach...

cwillisf replied:
The problem I ran into is that variable serialization wants to call `target.lookupVariableByNameAndType(...)` in order to get the variable's name and type, so I needed to feed `target` all the way down the line. That's a bit awkward since the "target" being passed into `serializeBlocks` here is the flattened target, not the "real" target, so variable serialization does `target = this.runtime.getTargetById(target.id);` to get the real target before calling `target.lookupVariableByNameAndType`.

Given that, I started thinking we should just pass down the target ID instead of the flattened target object. Then I started thinking about passing a "context" object in case we need more context info in the future instead of just the target ID. Then I was like, "well, the flattened target object is kind of the whole context" -- which is how I got here. Maybe it would be better to just pass in the target ID instead, if for no other reason than to avoid hassling extension authors with flattened targets.

kchadha replied:
Agreed about hassling extension authors with flattened targets. I'm okay with passing down a target ID but that also assumes that everywhere the target will be used (currently only in the custom serialization code) there will be access to the runtime. Maybe we can just pass around the hydrated target and call it `runtimeTarget` or something to indicate that it's not flattened? 

Related to a different suggestion I made, however, it seems like we might also need to start passing around the `runtime` as well, in which case `runtime + targetID` is okay too.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -791,30 +819,58 @@ const deserializeFields = function (fields) {
     return obj;
 };
 
+const decodeBlock = function (block) {
```
Maybe this can be called `decodeBlockOpcode` to be a little more descriptive (and set it apart from `deserializeBlock`. There's also a function exported by this module that has somewhat similar functionality (`getExtensionIdForOpcode`) with one key difference -- it checks that the `extensionId` is in the `CORE_EXTENSIONS` list defined in this file.

We might want to modify `getExtensionIdForOpcode` to use this new function, or somehow combine/consolidate the overlapping parts of the two functions.

cwillisf replied:
Notes from discussion: In the long run it seems likely that we'll need to do a two-pass approach to loading. In the first pass we'll initiate loading each extension that the project uses, then we'll wait for them all to finish loading, then we'll do a second pass to actually load the blocks. We don't need that just yet since internal extensions can load synchronously, but we'll need it once we start loading extensions asynchronously.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -791,30 +819,58 @@ const deserializeFields = function (fields) {
         }
         const block = blocks[blockId];
+        const {extensionId, extendedOpcode} = decodeBlock(block);
+        const extensionInfo = runtime.getExtensionInfo(extensionId);
```
Hmmmm I don't think this will work if the runtime doesn't already have the appropriate extension loaded.

In the future if we have non-core extensions with blocks with custom serialization, we'll have to figure something else out...

This isn't a problem for core extensions because we load them when we initialize the VM.

kchadha replied:
I remember we talked about this before but I forget if we had a solution for the future problem or not.

cwillisf replied:
Notes from discussion: In the long run it seems likely that we'll need to do a two-pass approach to loading. In the first pass we'll initiate loading each extension that the project uses, then we'll wait for them all to finish loading, then we'll do a second pass to actually load the blocks. We don't need that just yet since internal extensions can load synchronously, but we'll need it once we start loading extensions asynchronously.

kchadha replied:
To clarify part of the note above:

> We don't need that just yet since internal extensions can load synchronously, but we'll need it once we start loading extensions asynchronously.

We don't need that just yet since _the extensions for core block packages get loaded synchronously when the VM is initialized_, but we'll need it _for non-core extensions that start having custom serialization_.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -791,30 +819,58 @@ const deserializeFields = function (fields) {
+        if (customDeserialize) {
+            const newBlock = customDeserialize(block);
+            newBlock.id = uid();
+            delete blocks[blockId];
```
I think we should talk about whether this line is needed or not. There is an analogous line below (https://github.com/LLK/scratch-vm/pull/2279/files?file-filters%5B%5D=.js#diff-35528b8c179a308ffa56ce087ed40052R868) because in the old way of serializing primitives, we are not serializing block id information. I don't know that custom serialization/deserialization will necessarily be similar in that way where we won't be serializing an ID as a part of the custom serialization.

kchadha commented on src/blocks/data.js:
```diff
@@ -759,6 +769,40 @@ class Scratch3DataBlocks {
+            VARIABLE: {
+                name: 'VARIABLE',
+                value: primitiveArray[1],
+                id: primitiveArray[2],
```
Hmmm I think this `deserializeVariable` code should be producing the new dynamic block json as opposed to the old variable code block json... (same with the hardcoded line `result.opcode = 'data_variable';` above).

I believe we want the `deserializeVariable` code to be creating a block with a mutation with the blockInfo object in it instead...

kchadha reviewed (changes requested):
I did a preliminary review of this code. I tried to be pretty thorough in this pass, but did not look carefully at the documentation or the new `_prepareSerializationInfo` function yet. I think there's a lot to talk about here, so it might be good to discuss some of this in person.

Commit:
address code review feedback: smaller items

- Fix a lint error due to an unused constant.
- Don't return an extended opcode from `deserializeVariable` (more
  changes needed here).
- If `refreshBlocks` encounters an extension ID mismatch, keep the old
  ID instead of switching to the new one.
- Rename `decodeBlock` to `getExtensionAndOpcode` and clarify its
  purpose. In particular, distinguish it from `getExtensionIdForOpcode`.

Commit:
Merge branch 'e16n' into e16n-serialization-conservative

Commit:
simplify target access in custom serialization

Commit:
avoid storing extensionInfo on blocks

Commit:
in custom deserialize path, support dynamic blockInfo mutation

Commit:
fix unit tests

Commit:
fix integration tests

kchadha commented on src/blocks/data.js:
```diff
@@ -768,6 +777,62 @@ class Scratch3DataBlocks {
+        const result = {
+            opcode: 'variable',
+            fields: {
+                VARIABLE: {
```
I don't think we need this field descriptor here. This block should only need the mutation.

kchadha commented on src/blocks/data.js:
```diff
@@ -768,6 +777,62 @@ class Scratch3DataBlocks {
+        // something that looks like the block's entry in `getInfo()`. If we do either of those things we'll need to
+        // decide how we want to handle the block's (x,y) position and `topLevel` flag.
+        const result = {
+            opcode: 'variable',
```
This opcode is incorrect, the top level block opcode is the "extended" opcode.

cwillisf replied:
I was trying to avoid having the extension every mention the extended opcode, since in theory we might sometimes post-process the extension ID so that it isn't what the extension author expects. Instead I fix up the ID after the fact, in the function which calls this one.

kchadha commented on src/blocks/data.js:
```diff
@@ -768,6 +777,62 @@ class Scratch3DataBlocks {
+            parent: null,
+            shadow: false,
+            mutation: {
+                blockInfo: {
```
Similar to the comment above, I think it would be better for this function to just return this `blockInfo` object and have the rest of this top-level object constructed in either a helper function or by the sb3 deserialize code itself.

kchadha commented on src/blocks/data.js:
```diff
@@ -768,6 +777,62 @@ class Scratch3DataBlocks {
         this.runtime.requestToolboxExtensionsUpdate();
     }
+
+    serializeVariable (block, target) {
+        const variableName = block.mutation.blockInfo.text;
```
Probably this function should just take `blockInfo` as an argument, and then some helper function outside of the extension definition should handle high level block things like setting the x and y.

cwillisf replied:
The problem is that variables have their own custom way of serializing x and y, so it kind of needs to be here. That's also why the custom deserialize function needs to return the whole block instead of just the blockInfo guts :/

kchadha commented on src/engine/mutation-adapter.js:
```diff
@@ -34,7 +34,7 @@ const mutatorTagToObject = function (dom) {
  * @return {object} Object representing the mutation.
  */
-const mutationAdpater = function (mutation) {
+const mutationAdapter = function (mutation) {
```
😅

kchadha commented on src/extension-support/extension-manager.js:
```diff
@@ -171,16 +155,33 @@ class ExtensionManager {
+        const allPromises = [];
+        loadedExtensions.forEach((extensionInfo, oldId) => {
+            const serviceName = extensionInfo.serviceName;
+            allPromises.push(dispatch.call(serviceName, 'getInfo')
```
I think the `dispatch.call` here should be on its own line for a little bit more clarity when reading.

kchadha commented on src/extension-support/extension-manager.js:
```diff
@@ -389,6 +393,34 @@ class ExtensionManager {
+            const functionNames = rawSerializationInfo[opcode];
+            const extendedOpcode = `${extensionId}_${opcode}`;
+            const boundFunctions = {};
+            if (functionNames.serialize) {
```
Do we want to throw an error or log a warning if `serialize` or `deserialize` are not specified in the serialize descriptor for the opcode?

kchadha commented on src/serialization/sb3.js:
```diff
@@ -95,16 +97,59 @@ const primitiveOpcodeInfoMap = {
     data_listcontents: [LIST_PRIMITIVE, 'LIST']
 };
 
+const primitiveIdMap = (() => {
```
Could you add a comment for this function?

kchadha commented on src/serialization/sb3.js:
```diff
@@ -95,16 +97,59 @@ const primitiveOpcodeInfoMap = {
+    if (categoryId) {
+        return runtime.getExtensionInfo(categoryId);
+    }
+    // undefined
```
cruft?

kchadha replied:
Should we log a warning here if the categoryId was not found?

cwillisf replied:
This can happen legitimately for now (until 100% extensionification) so I added some more comment here to explain that.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -272,6 +319,28 @@ const compressInputTree = function (block, blocks) {
+    // note: if we ever need the simple / non-extended opcode, it's just extendedOpcode.substring(firstSplit + 1)
+    return {
+        extendedOpcode,
+        extensionId: extendedOpcode.substring(0, firstSplit)
```
Elsewhere we use `StringUtil.splitFirst` to get this info, might be good to use that here for consistency.

kchadha commented on src/serialization/sb3.js:
```diff
@@ -288,20 +357,20 @@ const getExtensionIdForOpcode = function (opcode) {
- * in the serialized blocks.
+ * @param {RenderedTarget} target The target containing the blocks to be serialized.
+ * @param {Runtime} runtime The VM runtime
+ * @param {Set} extensions A set of extensions to add extension IDs to
```
This param comment doesn't really explain which extension IDs are being tracked or why

kchadha commented on src/serialization/sb3.js:
```diff
@@ -434,33 +503,33 @@ const serializeComments = function (comments) {
- * @param {object} target The target to be serialized.
+ * @param {RenderedTarget} target The target to be serialized.
+ * @param {Runtime} runtime The VM runtime
  * @param {Set} extensions A set of extensions to add extension IDs to
```
I guess we have the same problem here.. so you can probably ignore the comment above.

Alternatively we can make it more descriptive -- the set of extension IDs from the blocks being used by this target (or some better phrasing)

kchadha commented on src/serialization/sb3.js:
```diff
@@ -792,28 +855,57 @@ const deserializeFields = function (fields) {
- * Covnert serialized INPUT and FIELD primitives back to hydrated block templates.
+ * Convert dynamic blockInfo JSON from a custom deserialization function into an XML element ready for runtime use.
+ * @param {object} dynamicBlockInfo - a BlockInfo-like object with properties like `opcode` and `isDynamic`.
+ * @returns {Element} - an XML Element with tag name 'mutation' and an attribute called 'blockInfo'
```
The `mutationAdapter` function does not return an XML Element, it takes an XML element (or string) and returns a json object w/ properties relevant to the VM.

cwillisf replied:
oops - yeah, I forgot to update the comment after settling on how this should work...

kchadha commented on src/serialization/sb3.js:
```diff
@@ -792,28 +855,57 @@ const deserializeFields = function (fields) {
+        if (customDeserialize) {
+            block = customDeserialize(block);
+            blocks[blockId] = block; // customDeserialize might have made a new object; make sure to keep it!
+            // the deserialize function should return opcode 'foo' and we'll fix it to 'ext_foo'
```
Related to the comment earlier in this PR, if we're doing fix-up here anyways of what the customDeserialize function returns, I think we might as well construct the block JSON here instead of the customDeserialize function and have that function just return the blockInfo.

kchadha reviewed (changes requested):
Left comments in-line in the PR.

Commit:
address smaller items from code review feedback

kchadha commented:
Unassigning myself from this PR since we discussed offline (a few weeks ago) that maybe custom serialization in the extension spec is not the right approach since we only want it in this specific case of the core variables extension. We can handle this special case in the "backend" (e.g. vm runtime or extension manager) instead of making this a developer facing feature. We also discussed removing this from the immediate scope of this "first phase" of the e16n project (wrapping up by the end of this year).

towerofnix commented:
>We can handle this special case in the "backend" (e.g. vm runtime or extension manager) instead of making this a developer facing feature. We also discussed removing this from the immediate scope of this "first phase" of the e16n project (wrapping up by the end of this year).

does this mean custom variable types are out of the question for developer extensions? not that I actually know what's going on in this code, nor that it'd mean that wouldn't be supported *ever*, but this was one of the most exciting things about extensions IMO (localstorage-based variables, etc).
